Pollutant Mods

abhiramm7 opened the pull request:
*This work is in collaboration with @bemason. We also want to acknowledge the help of @jennwuu,  @katmratliff, and @dickinsonre* :)

Adds setters for pollutants in nodes and links to enable modeling of pollutant transformation using external tools like pyswmm.

New Functionality:

1. Getters for node inflow pollutant, reactor concentration (c2 in qualrout.c *findStorageQual* function call), and HRT
2. Setters for pollutants in nodes and links

Issues:
Mass balance error seems to be more robust for nodes than links. As of now, I have a higher water quality error tolerance for links (5%)  than nodes (1%) in the implemented tests.

michaeltryby commented:
Hey @bemason and @abhiramm7, Let me start by saying this is a great idea! Exposing pollutants in the Toolkit API will allow SWMM modelers to bypass the solver's water quality limitations much the same as the Toolkit allowed @bemcdonnell to bypass the limitations of the controls parser. 

Do you have a write up of your use case? A writeup would be nice so that myself and others can better understand what you are doing.  

I have looked over your PR and it looks pretty good. In it you introduce the concept of an external pollutant. What services does SWMM perform for external pollutants -- routing, mass balance, anything else? Would it simplify things to assume pollutants are either all internal or all external? 

Thanks for your work on this! I look forward to your response.

abhiramm7 commented:
Hi, 

Thanks for your comments, @michaeltryby. As you mentioned, our idea is to extend the water quality modeling capabilities of SWMM by providing an API that enables users to inject and modify the pollutant in SWMM. We believe that this approach might be sustainable in the long run than augmenting the pollutant treatment models provided in SWMM. Through this API, users can read the inflow pollutant concentration into a node (or a link), model the pollutant transformation in a high-level programming language like Python, and set the computed values back into the SWMM. This approach relies upon the existing quality routing in SWMM's engine for transporting the pollutants through the stormwater network. 

Modeling pollutant transformations in Python enables us to leverage the scientific computing stack to develop complex pollutant models. We plan on releasing a Python package (StormReactor) that builds on the getters and setters in this PR to provide an easy to use toolkit for modeling pollutant transformations. StormReactor will provide implementations of CSTR, Plug flow reactors, and other established pollutant models for users to build on. We also plan on submitting a PR to PySWMM with the pollutant getter and setters. StormReactor would then have PySWMM as a dependency. 

In terms of specific use cases, we have updated the pollutant implementation in SWMM such that users can introduce pollutants in links to model flow driven erosion in channels. In nodes, as alluded above, users can model pollutant removal as a combination of reactors; this would vital in pollutant based control of stormwater systems. 

In terms of implementation details, we introduced "extPollutFlag" in nodes and links. This flag is pollutant and object (i.e., link or node) specific. It enables users to control which pollutant they want to model externally—this flag control when the SWMM skips its pollutant removal in favor of external pollutant values. Externally set pollutant values are accounted for in the SWMM's mass balance functionality. In this implementation, all the pollutants can be considered as internal, which can be modified externally. 

I hope this provides a better outline of the PR. Do let us know if you have any other questions.

Thanks,
Abhi

abhiramm7 commented:
@michaeltryby have you had a chance to look at the PR summary? Is there anything else you would recommend that we incorporate into this PR?  

@cbuahin somehow mac testing seems to stuck. I was unable to pinpoint what was causing the error. I was able run the tests on the my local machine. When you get a moment can give this a look?

bemason commented:
Hi @michaeltryby, @katmratliff, @bemcdonnell , and @cbuahin, I am hoping to get the paper submitted on the toolbox that uses these changes in the next two weeks. Do you have an estimate on when these changes will be reviewed? I believe I need these changes to be accepted before I submit the paper. Thank you!

bemcdonnell commented:
@bemason, would you be totally opposed if we integrate this into PySWMM-v1.0?  That is just a couple weeks out now.  Else, we can do a pyswmm release of v0.7.0 with this work embedded.  The different is a pretty big shift in pyswmm architecture that will be more friendly x-platform x-architecture.

bemason commented:
@bemcdonnell v1.0 is fine for PySWMM as long as these SWMM edits are merged in the next week or so.

bemcdonnell commented:
@bemason I think the plan is to finish and merge changes for the beta release of pyswmm-v1.0 next week. Then there should be a few iterations of changes/bug fixes.  After that, the contributions made in this pull request will have to be rebased and the tests will all have to pass (which might request a little effort from @abhiramm7 - he can let us know).  Once all is well, we can merge here and then we can stage a release of pyswmm (probably v1.1).  There are going to be several quick things happening next week.  Sorry for the delay here.  All of a sudden these projects have a lot of horsepower

bemason commented:
Hi All, Just checking on the timeline for this pull request. Thanks!

bemcdonnell commented:
@jennwuu Could you help come up with a plan to bring these new features into the next release?  There have been some fundamental changes in the coded workflow thus we will have pretty substantial conflicts here
